File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> strMap;
        int maxLen = 0;
        int left = 0;
        int right = 0;

        while(right < s.length()){
            if(strMap.find(s[right]) != strMap.end() && strMap[s[right]] >= left){
                left = strMap[s[right]]+1;
            }
            maxLen = max(maxLen, right-left+1);
            strMap[s[right]] = right;
            right++;
        }

        return maxLen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest_substr = set()
        l, r = 0, 0
        max_len = 0
        
        while r < len(s):
            while s[r] in longest_substr:
                longest_substr.remove(s[l])
                l += 1

            max_len = max(max_len, r-l+1)
            longest_substr.add(s[r])
            r += 1
        
        return max_len